### PR TITLE
Change == implementation and implement ===

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
This PR 
- updates `==` to perform a deep value equality check
- implements `===` as a reference equality check

This PR also remove some invalid instructions and equality checks that were previously allowed.

#### TODOs (18/10/2018)
- implement `==` for tables
- validate that `==`, `!=` is only used with `anydata`

### CODE REVIEW

File: cli/ballerina-cli-utils/src/main/ballerina/packaging_search/packaging_search.bal [hunk 1]
```diff
@@ -33,7 +33,7 @@ function search (http:Client definedEndpoint, string url, string querySearched,
     } else {
         jsonResponse = check (httpResponse.getJsonPayload());
         json[] artifacts = check <json[]> jsonResponse.artifacts;
-        if (artifacts == null || lengthof artifacts > 0) {
+        if (lengthof artifacts > 0) {
```
Comment (vinok88) [inline cli/ballerina-cli-utils/src/main/ballerina/packaging_search/packaging_search.bal]: What does this have to do with equality check?

Comment (NatashaWso2) [reply cli/ballerina-cli-utils/src/main/ballerina/packaging_search/packaging_search.bal]: Here artifacts can never be null. If there are no packages to be found we simply return an artifact array of zero. Since we already check if the artifact array size is greater than zero it is okay to remove this check

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/equal_and_not_equal_operation.bal [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+    OpenEmployee e4 = { name: "Em", area: 51 };
+
+    OpenEmployee e5 = { name: "Em" };
+    OpenPerson e6 = { name: "Em" };
+
+    OpenEmployee e7 = { name: "Em", id: 4000, dept: "finance" };
+    OpenEmployee e8 = { name: "Em", id: 4000, dept: "hr" };
+
+    return (e1 == e2) || !(e1 != e2) || (e3 == e4) || !(e3 != e4) || isEqual(e5, e6) ||
+        (e7 == e8) || !(e7 != e8);
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/equal_and_not_equal_operation.bal]: Shall we have return statement in one line?

File: cli/ballerina-cli-utils/src/main/ballerina/packaging_search/packaging_search.bal (hunk 1 above)
Comment (MaryamZi) [reply cli/ballerina-cli-utils/src/main/ballerina/packaging_search/packaging_search.bal]: Yeah, basically an array cannot be `null` (default is an empty array), but we had allowed this equality check previously. That equality check is no longer allowed, thus this change.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java [hunk 3]
```diff
@@ -69,10 +74,28 @@
...
+                                   "org.apache.axiom.c14n.impl.Canonicalizer20010315OmitComments");
+            Canonicalizer.register("http://www.w3.org/TR/2001/REC-xml-c14n-20010315#WithComments",
+                                   "org.apache.axiom.c14n.impl.Canonicalizer20010315WithComments");
+            Canonicalizer.register("http://www.w3.org/2001/10/xml-exc-c14n#",
+                                   "org.apache.axiom.c14n.impl.Canonicalizer20010315ExclOmitComments");
+            Canonicalizer.register("http://www.w3.org/2001/10/xml-exc-c14n#WithComments",
+                                   "org.apache.axiom.c14n.impl.Canonicalizer20010315ExclWithComments");
+            canonicalizer = Canonicalizer.getInstance("http://www.w3.org/2001/10/xml-exc-c14n#WithComments");
+        } catch (InvalidCanonicalizerException | AlgorithmAlreadyRegisteredException e) {
+            throw new BallerinaException("Error initializing canonicalizer " + e.getMessage());
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java]: add a colon to merge the messages :)

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java [hunk 4]
```diff
@@ -365,6 +388,54 @@ public static BValue convertToJSON(BXML xml, String attributePrefix, boolean pre
...
+     *
+     * @param xmlOne    the first XML value
+     * @param xmlTwo    the second XML value
+     * @return  true if the two are equal, false if not
+     */
+    public static boolean isEqual(BXML<?> xmlOne, BXML<?> xmlTwo) {
+        XMLNodeType xmlOneNodeType = xmlOne.getNodeType();
+        XMLNodeType xmlTwoNodeType = xmlTwo.getNodeType();
+
+        if (xmlOneNodeType == XMLNodeType.SEQUENCE && xmlTwoNodeType == XMLNodeType.SEQUENCE) {
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/XMLUtils.java]: There can be a situation where one is a sequence of size 1 (singleton, a sequence with a single element type xml in it), and the other is an element type xml (singleton, but not a sequence). We need to treat both of them as singletons.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 5]
```diff
@@ -1427,46 +1427,207 @@ public boolean isAssignableToFiniteType(BType type,
...
+                case TypeTags.RECORD:
+                    for (BType rhsMemberType : rhsTypes) {
+                        if (checkStructEquivalency(rhsMemberType, lhsMemberType) ||
+                                checkStructEquivalency(lhsMemberType, rhsMemberType)) {
+                            // intersection exists if one is assignable to the other
+                            matchFound = true;
+                            break;
+                        }
+                    }
+                    break;
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: we can stream and do  anyMatch(). It will simplify the code.

Can return if a match is found, then and there.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Refactored the method and addressed all suggestions, thanks!
